### ISSUE TITLE
Unsupported dynamic function invocation

### ISSUE THREAD

**Issue by glwagner** [climate-machine/Oceananigans.jl#248]
I'm getting an error when trying to compile constant Smagorinsky:

```julia
ERROR: LoadError: InvalidIRError: compiling #12(RegularCartesianGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Flo
at64},Base.TwicePrecision{Float64}}}, PlanetaryConstants{Float64}, LinearEquationOfState{Float64}, ConstantSmagorinsky{Fl
oat64}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global
}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CU
DAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnat
ive.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.C
uDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDevi
ceArray{Float64,3,CUDAnative.AS.Global}, Forcing{typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func),typeof(Oce
ananigans.zero_func),typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func)}) resulted in invalid LLVM IR
Reason: unsupported call to the Julia runtime (call to jl_f__apply)
Stacktrace:
 [1] overdub at /data5/glwagner/.julia/packages/Cassette/xggAf/src/context.jl:260
 [2] ν_ccc at /data5/glwagner/Projects/Oceananigans.jl/src/closures/constant_smagorinsky.jl:109
 [3] ν_Σᵢⱼ at /data5/glwagner/Projects/Oceananigans.jl/src/closures/closure_operators.jl:405
 [4] ∂x_faa at /data5/glwagner/Projects/Oceananigans.jl/src/closures/closure_operators.jl:64
 [5] ∂x_2ν_Σ₁₁ at /data5/glwagner/Projects/Oceananigans.jl/src/closures/closure_operators.jl:409
 [6] ∂ⱼ_2ν_Σ₁ⱼ at /data5/glwagner/Projects/Oceananigans.jl/src/closures/closure_operators.jl:432
 [7] calculate_interior_source_terms! at /data5/glwagner/Projects/Oceananigans.jl/src/time_steppers.jl:152
 [8] #12 at /data5/glwagner/.julia/packages/GPUifyLoops/hBRid/src/context.jl:136
Reason: unsupported dynamic function invocation (call to Cassette.overdub)
```

I think this is specific to the package upgrades. I also got this error when running the `rayleigh_benard_passive_tracer.jl` script (associated with the function that forces salinity). Mysteriously, it does not throw this error when it runs the tests. Not sure where to raise an issue (if this is an issue)...

_Originally posted by @glwagner in https://github.com/climate-machine/Oceananigans.jl/pull/245#issuecomment-496472606_

**Comment by glwagner** [comment]
This kind of error arises when using the environment on the branch `integrate_turbulence_closures`:

```julia
(Oceananigans) pkg> st
Project Oceananigans v0.5.4
    Status `/archive1/glwagner/Projects/Oceananigans.jl/Project.toml`
  [79e6a3ab] Adapt v0.4.2
  [c5f51814] CUDAdrv v3.0.1
  [be33ccc6] CUDAnative v2.1.1
  [3a865a2d] CuArrays v1.0.2
  [7a1cc6ca] FFTW v0.2.4
  [ba82f77b] GPUifyLoops v0.2.3
  [4138dd39] JLD v0.9.1
  [30363a11] NetCDF v0.7.3
  [90137ffa] StaticArrays v0.10.3
  [a759f4b9] TimerOutputs v0.5.0
  [8ba89e20] Distributed 
  [37e2e46d] LinearAlgebra 
  [de0858da] Printf 
  [9a3f8284] Random 
  [10745b16] Statistics 
  [8dfed614] Test 
```

**Comment by glwagner** [comment]
This may not be an issue for `Oceananigans`, but it could be good to discuss here first.

**Comment by ali-ramadhan** [comment]
I had issues with GPUifyLoops v0.2.4: GPU tests were crashing with similar `unsupported dynamic function invocation (call to Cassette.overdub)` errors. So I downgraded to v0.2.3. See: https://github.com/vchuravy/GPUifyLoops.jl/issues/66

I might try v0.2.1 but not sure if that'll help as not much changed between v0.2.1 and v0.2.3. Anything lower than that won't work or will be super slow to compile.

@vchuravy points out that it may not be a GPUifyLoops error. I'll have a closer look at the lines the stacktrace is pointing at.

**Comment by ali-ramadhan** [comment]
Another idea: We can go back to the older `Project.toml` and try running this again in case it's a Julia CUDA issue.

**Comment by ali-ramadhan** [comment]
Also, which commit are you running this from? Latest commit on your `modified-constant-smag` branch?

**Comment by glwagner** [comment]
I did another test from `integrate-turbulence-closures`.

I just tried downgrading `GPUifyLoops` to `v0.2.1` and got the same error.

**Comment by glwagner** [comment]
a 'simple forcing' script to reproduce this error is here:

https://github.com/climate-machine/Oceananigans.jl/blob/631b861bcfdbfd894c4cb6b2257a16aa8c55f8cc/sandbox/simple_forcing.jl

**Comment by glwagner** [comment]
Suggestions as how to downgrade the CUDA packages?

**Comment by ali-ramadhan** [comment]
Hmmm, maybe try going back to the Project.toml and Manifest.toml files from before PR #244 

https://github.com/climate-machine/Oceananigans.jl/blob/1bd203648dd9070574fa08568d1b68bea4fe0590/Project.toml

https://github.com/climate-machine/Oceananigans.jl/blob/1bd203648dd9070574fa08568d1b68bea4fe0590/Manifest.toml

Could do it manually but I found it a pain to figure out exactly which combination of package versions would actually work, and we're restricted to recent versions of most CUDA packages as kernel compile times were recently improved.

**Comment by ali-ramadhan** [comment]
Still trying to figure out which commit/branch to look at. The stacktrace mentions `constant_smagorinsky.jl:109` but I can't find a `constant_smagorinsky.jl` file with more than 100 lines.

**Comment by glwagner** [comment]
using the script 

https://github.com/climate-machine/Oceananigans.jl/blob/631b861bcfdbfd894c4cb6b2257a16aa8c55f8cc/sandbox/simple_forcing.jl

on the branch `integrate-turbulence-closures` that is part of PR #245 , I get the following error:

```
julia> include("simple_forcing.jl")
[ Info: Precompiling Oceananigans [9e8cae18-63c1-5223-a75c-80ca9d6e9a09]
WARNING: Method definition recurse(Cassette.Context{N, M, T, P, B, H} where H<:Union{Cassette.DisableHooks, Nothing} where B<:Union{Nothing, Base.IdDict{Module, Base.Dict{Symbol, Cassette.BindingMeta}}} where P<:Cassette.AbstractPass where T<:Union{Nothing, Cassette.Tag{N, X, E} where E where X where N<:Cassette.AbstractContextName} where M where N<:Cassette.AbstractContextName, Any...) in module Cassette at /data5/glwagner/.julia/packages/Cassette/xggAf/src/overdub.jl:521 overwritten in module GPUifyLoops at /data5/glwagner/.julia/packages/Cassette/xggAf/src/overdub.jl:521.
WARNING: Method definition overdub(Cassette.Context{N, M, T, P, B, H} where H<:Union{Cassette.DisableHooks, Nothing} where B<:Union{Nothing, Base.IdDict{Module, Base.Dict{Symbol, Cassette.BindingMeta}}} where P<:Cassette.AbstractPass where T<:Union{Nothing, Cassette.Tag{N, X, E} where E where X where N<:Cassette.AbstractContextName} where M where N<:Cassette.AbstractContextName, Any...) in module Cassette at /data5/glwagner/.julia/packages/Cassette/xggAf/src/overdub.jl:508 overwritten in module GPUifyLoops at /data5/glwagner/.julia/packages/Cassette/xggAf/src/overdub.jl:508.

CUDA-enabled GPU(s) detected:
CuDevice(0): Quadro P6000
[ Info: Building the CUDAnative run-time library for your sm_61 device, this might take a while...
ERROR: LoadError: InvalidIRError: compiling #12(RegularCartesianGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}, PlanetaryConstants{Float64}, LinearEquationOfState{Float64}, Oceananigans.TurbulenceClosures.ConstantAnisotropicDiffusivity{Float64}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, Forcing{typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func),typeof(FS)}) resulted in invalid LLVM IR
Reason: unsupported dynamic function invocation (call to Cassette.overdub)
Stacktrace:
 [1] S★ at /archive1/glwagner/Projects/Oceananigans.jl/sandbox/simple_forcing.jl:21
 [2] FS at /archive1/glwagner/Projects/Oceananigans.jl/sandbox/simple_forcing.jl:22
 [3] calculate_interior_source_terms! at /archive1/glwagner/Projects/Oceananigans.jl/src/time_steppers.jl:178
 [4] #12 at /data5/glwagner/.julia/packages/GPUifyLoops/hBRid/src/context.jl:136
Reason: unsupported dynamic function invocation (call to Cassette.overdub)
Stacktrace:
 [1] FS at /archive1/glwagner/Projects/Oceananigans.jl/sandbox/simple_forcing.jl:22
 [2] calculate_interior_source_terms! at /archive1/glwagner/Projects/Oceananigans.jl/src/time_steppers.jl:178
 [3] #12 at /data5/glwagner/.julia/packages/GPUifyLoops/hBRid/src/context.jl:136
Reason: unsupported dynamic function invocation (call to Cassette.overdub)
Stacktrace:
 [1] calculate_interior_source_terms! at /archive1/glwagner/Projects/Oceananigans.jl/src/time_steppers.jl:178
 [2] #12 at /data5/glwagner/.julia/packages/GPUifyLoops/hBRid/src/context.jl:136
Stacktrace:
 [1] check_ir(::CUDAnative.CompilerJob, ::LLVM.Module) at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/compiler/validation.jl:114
 [2] macro expansion at /data5/glwagner/.julia/packages/TimerOutputs/7zSea/src/TimerOutput.jl:216 [inlined]
 [3] #codegen#119(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/compiler/driver.jl:186
 [4] #codegen at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/compiler/driver.jl:0 [inlined]
 [5] #compile#118(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/compiler/driver.jl:47
 [6] #compile#117 at ./none:0 [inlined]
 [7] compile at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/compiler/driver.jl:28 [inlined] (repeats 2 times)
 [8] macro expansion at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/execution.jl:378 [inlined]
 [9] #cufunction#159(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative.cufunction), ::getfield(GPUifyLoops, Symbol("##12#13")){typeof(Oceananigans.calculate_interior_source_terms!)}, ::Type{Tuple{RegularCartesianGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},PlanetaryConstants{Float64},LinearEquationOfState{Float64},Oceananigans.TurbulenceClosures.ConstantAnisotropicDiffusivity{Float64},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},Forcing{typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func),typeof(Oceananigans.zero_func),typeof(FS)}}}) at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/execution.jl:347
 [10] cufunction(::Function, ::Type) at /data5/glwagner/.julia/packages/CUDAnative/9rZcJ/src/execution.jl:347
 [11] macro expansion at /data5/glwagner/.julia/packages/GPUifyLoops/hBRid/src/GPUifyLoops.jl:113 [inlined]
 [12] macro expansion at ./gcutils.jl:87 [inlined]
 [13] #launch#46(::Base.Iterators.Pairs{Symbol,Tuple{Int64,Int64,Vararg{Int64,N} where N},Tuple{Symbol,Symbol},NamedTuple{(:threads, :blocks),Tuple{Tuple{Int64,Int64},Tuple{Int64,Int64,Int64}}}}, ::Function, ::GPUifyLoops.CUDA, ::typeof(Oceananigans.calculate_interior_source_terms!), ::RegularCartesianGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}, ::Vararg{Any,N} where N) at /data5/glwagner/.julia/packages/GPUifyLoops/hBRid/src/GPUifyLoops.jl:110
 [14] #launch at ./none:0 [inlined]
 [15] macro expansion at /data5/glwagner/.julia/packages/GPUifyLoops/hBRid/src/GPUifyLoops.jl:54 [inlined]
 [16] time_step!(::Model{GPU,RegularCartesianGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},Oceananigans.TurbulenceClosures.ConstantAnisotropicDiffusivity{Float64},Float64}, ::Int64, ::Float64) at /archive1/glwagner/Projects/Oceananigans.jl/src/time_steppers.jl:74
 [17] top-level scope at util.jl:156
 [18] include at ./boot.jl:326 [inlined]
 [19] include_relative(::Module, ::String) at ./loading.jl:1038
 [20] include(::Module, ::String) at ./sysimg.jl:29
 [21] include(::String) at ./client.jl:403
 [22] top-level scope at none:0
in expression starting at /archive1/glwagner/Projects/Oceananigans.jl/sandbox/simple_forcing.jl:35
```

**Comment by glwagner** [comment]
@ali-ramadhan that error was produced on a branch on my fork called `modified-constant-smag`, from this file:

https://github.com/glwagner/Oceananigans.jl/blob/modified-constant-smag/src/closures/constant_smagorinsky.jl

**Comment by vchuravy** [comment]
You can try https://github.com/JuliaGPU/CUDAnative.jl/pull/404 to see if you get a better error

**Comment by glwagner** [comment]
@vchuravy solved it (naturally)! The issue is that functions 

1. cannot reference non-`const` global variables (as was occurring in `simple_forcing.jl`
2. cannot have a `Union` return type, as is the case in many operators because of various instances of branches that `0` and `0.5` (when the other branch returns `Float32` or `Float64`)

**Comment by ali-ramadhan** [comment]
That's good to know. Yeah, we definitely need to clean up some of these type instabilities.

Do you want to close this issue?

**Comment by glwagner** [comment]
Resolved by #245.
